File: src/daily_value.py
"""Daily highlights across stored canonical picks; never change pick direction.

This is a heuristic risk limit, not an empirically optimized prediction model.
Keep the unrounded JSON contract in sync with the frontend daily-value policy.
"""
from __future__ import annotations

from datetime import datetime, timedelta, timezone
import math
import re


POLICY_VERSION = "daily-value-v1"
MIN_HIT = 0.50
MIN_RETURN = -0.15
BASE_PER_LEAGUE = 3
EPSILON = 1e-12
KST = timezone(timedelta(hours=9))


def _finite(value: object) -> float | None:
    # Accept numeric strings from older artifacts, but not booleans/containers.
    if isinstance(value, bool) or not isinstance(value, (int, float, str)):
        return None
    if isinstance(value, str) and not re.fullmatch(
        r"[+-]?(?:[0-9]+\.?[0-9]*|\.[0-9]+)(?:[eE][+-]?[0-9]+)?", value.strip()
    ):
        return None
    try:
        number = float(value)
    except (ValueError, OverflowError):
        return None
    return number if math.isfinite(number) else None


def _probability(value: object) -> float | None:
    number = _finite(value)
    return number if number is not None and 0 < number < 1 else None


def _alias(candidate: dict, field: str, alias: str) -> object:
    value = candidate.get(field)
    return candidate.get(alias) if value is None else value


def daily_value_metrics(candidate: dict | None) -> dict:
    """Compute point and conservative returns, with strict validation markers."""
    candidate = candidate or {}
    metrics = {
        "policy_version": POLICY_VERSION,
        "probability": None,
        "comparison_probability": None,
        "break_even_probability": None,
        "expected_return": None,
        "comparison_return": None,
        "validated_probability": False,
        "validated_interval": False,
        "qualifies": False,
    }
    odds = _finite(_alias(candidate, "odds", "배당"))
    market = _probability(_alias(candidate, "market_prob", "시장확률"))
    if odds is None or odds <= 1 or market is None:
        return metrics

    final = _probability(candidate.get("predicted_hit_prob"))
    validated = (
        candidate.get("decision_pipeline_applied") is True
        and candidate.get("has_validated_edge") is True
        and final is not None
    )
    estimate = final if validated else market
    comparison = min(market, estimate)
    interval = candidate.get("probability_interval")
    trusted_interval = False
    if (
        validated
        and candidate.get("validated_uncertainty_available") is True
        and candidate.get("uncertainty_source") == "validated_residual_interval"
        and isinstance(interval, list)
        and len(interval) == 2
    ):
        lower, upper = map(_probability, interval)
        bound = _probability(candidate.get("probability_lower_bound"))
        trusted_interval = (
            lower is not None and upper is not None and bound is not None
            and lower <= estimate <= upper
            # The producer serializes the stated bound to four decimals while
            # retaining full precision for the interval endpoints.
            and abs(bound - lower) <= 5.0001e-5
        )
        if trusted_interval:
            comparison = lower

    expected_return = estimate * odds - 1
    comparison_return = comparison * odds - 1
    metrics.update({
        "probability": estimate,
        "comparison_probability": comparison,
        "break_even_probability": 1 / odds,
        "expected_return": expected_return,
        "comparison_return": comparison_return,
        "validated_probability": validated,
        "validated_interval": trusted_interval,
        "qualifies": (estimate >= MIN_HIT
                      and comparison_return >= MIN_RETURN - EPSILON),
    })
    return metrics


def _clean(value: object) -> str:
    return "" if value is None else str(value).strip()


def _date_parts(candidate: dict) -> tuple[str, str | None]:
    """ISO kickoff -> KST; never read the clock to infer a missing year."""
    try:
        kickoff = datetime.fromisoformat(_clean(candidate.get("kickoff_at")))
        if kickoff.tzinfo is None:
            kickoff = kickoff.replace(tzinfo=KST)
        day = kickoff.astimezone(KST).date().isoformat()
        return day[:4], day[5:]
    except (ValueError, OverflowError):
        match = re.match(r"^(\d{2})\.(\d{2})", _clean(candidate.get("date")))
        return _clean(candidate.get("year")), f"{match[1]}-{match[2]}" if match else None


def _day_league(candidate: dict, known_years: dict[str, set[str]]) -> tuple[str, str]:
    year, month_day = _date_parts(candidate)
    known = known_years.get(month_day, set())
    inferred = next(iter(known)) if len(known) == 1 else "undated"
    day = f"{year or inferred}-{month_day}" if month_day else "undated"
    return day, _clean(candidate.get("league")) or "리그 미분류"


def _rank(candidate: dict) -> tuple:
    metrics = candidate["daily_recommendation"]
    label = candidate.get("market_label")
    if label is None:
        label = candidate.get("label")
    key = "|".join(map(_clean, (
        candidate.get("round"), _alias(candidate, "game_no", "게임번호"),
        candidate.get("market"), label, _alias(candidate, "sel", "선택"),
    )))
    return (
        -metrics["comparison_return"], -metrics["expected_return"],
        -metrics["probability"],
        # JS lexical comparison uses UTF-16 code units, with no locale collation.
        _clean(candidate.get("kickoff_at") or candidate.get("date")).encode(
            "utf-16-be", errors="surrogatepass"),
        key.encode("utf-16-be", errors="surrogatepass"),
    )


def annotate_daily_values(candidates: list[dict]) -> list[dict]:
    """Return copies in input order, including every excluded canonical pick.

    Reason precedence: safety, invalid, hit_floor, return_floor, fallback,
    then rank/base/validated_extra. 'fallback' excludes a low-odds row when
    a qualifying primary exists; selected low-odds rows use 'base' or
    'validated_extra'. Ranks are one-based within the retained league pool.
    This pure full-list helper intentionally does not special-case started locks.
    """
    annotated = []
    groups: dict[tuple[str, str], list[dict]] = {}
    known_years: dict[str, set[str]] = {}
    for original in candidates:
        year, month_day = _date_parts(original or {})
        if year and month_day:
            known_years.setdefault(month_day, set()).add(year)
    for original in candidates:
        candidate = dict(original or {})
        metrics = daily_value_metrics(candidate)
        decision = {
            **metrics, "recommended": False, "league_rank": None,
            "reason_code": "rank",
        }
        candidate["daily_recommendation"] = decision
        annotated.append(candidate)
        odds = _finite(_alias(candidate, "odds", "배당"))
        if (
            _clean(candidate.get("market")) == "홀짝"
            or odds is None or odds <= 1 or odds >= 2.2
            or candidate.get("is_market_favorite") is False
            or candidate.get("final_reversal") is True
            or candidate.get("최종전환") is True
        ):
            decision["reason_code"] = "safety"
        elif metrics["probability"] is None:
            decision["reason_code"] = "invalid"
        elif metrics["probability"] < MIN_HIT:
            decision["reason_code"] = "hit_floor"
        elif not metrics["qualifies"]:
            decision["reason_code"] = "return_floor"
        else:
            groups.setdefault(_day_league(candidate, known_years), []).append(candidate)

    for rows in groups.values():
        primary = [row for row in rows if _finite(_alias(row, "odds", "배당")) >= 1.5]
        if primary:
            for row in rows:
                if _finite(_alias(row, "odds", "배당")) < 1.5:
                    row["daily_recommendation"]["reason_code"] = "fallback"
        for rank, row in enumerate(sorted(primary or rows, key=_rank), start=1):
            decision = row["daily_recommendation"]
            decision["league_rank"] = rank
            if rank <= BASE_PER_LEAGUE:
                decision.update(recommended=True, reason_code="base")
            elif decision["validated_interval"] and decision["comparison_return"] > 0:
                decision.update(recommended=True, reason_code="validated_extra")
    return annotated

File: web/src/lib/daily-value.js
import { eligibleFinalSelections, recommendationPriority } from "./recommendation-policy.js";

// Risk limits, not fitted probabilities or a claim of profitable performance.
export const POLICY_VERSION = "daily-value-v1";
export const MIN_HIT = 0.50;
export const MIN_RETURN = -0.15;
export const BASE_PER_LEAGUE = 3;

const number = (value) => {
  if (typeof value !== "number" && (typeof value !== "string" ||
      !/^[+-]?(?:\d+\.?\d*|\.\d+)(?:[eE][+-]?\d+)?$/.test(value.trim()))) return null;
  return Number.isFinite(Number(value)) ? Number(value) : null;
};
const probability = (value) => {
  const p = number(value);
  return p != null && p > 0 && p < 1 ? p : null;
};

/** Consume only normalized, current-revision probability metadata from the pipeline. */
export function dailyValueMetrics(candidate) {
  const invalid = {
    policy_version: POLICY_VERSION, probability: null, comparison_probability: null,
    break_even_probability: null, expected_return: null, comparison_return: null,
    validated_probability: false, validated_interval: false, qualifies: false,
  };
  const odds = number(candidate?.odds ?? candidate?.["배당"]);
  const market = probability(candidate?.market_prob ?? candidate?.["시장확률"]);
  if (odds == null || odds <= 1 || market == null) return invalid;
  const final = probability(candidate?.predicted_hit_prob);
  const validated = candidate?.decision_pipeline_applied === true &&
    candidate?.has_validated_edge === true && final != null;
  const estimate = validated ? final : market;
  const interval = candidate?.probability_interval;
  const lower = probability(candidate?.probability_lower_bound);
  const lo = Array.isArray(interval) && interval.length === 2 ? probability(interval[0]) : null;
  const hi = Array.isArray(interval) && interval.length === 2 ? probability(interval[1]) : null;
  const hasInterval = validated && candidate?.validated_uncertainty_available === true &&
    candidate?.uncertainty_source === "validated_residual_interval" &&
    lower != null && lo != null && hi != null && lo <= estimate && estimate <= hi &&
    // today_combo serializes the stated lower to four decimal places.
    Math.abs(lower - lo) <= 5.0001e-5;
  const comparison = hasInterval ? lo : Math.min(market, estimate);
  const expectedReturn = estimate * odds - 1;
  const comparisonReturn = comparison * odds - 1;
  return {
    policy_version: POLICY_VERSION, probability: estimate, comparison_probability: comparison,
    break_even_probability: 1 / odds, expected_return: expectedReturn,
    comparison_return: comparisonReturn, validated_probability: validated,
    validated_interval: hasInterval,
    qualifies: estimate >= MIN_HIT && comparisonReturn >= MIN_RETURN - 1e-12,
  };
}

const clean = (value) => String(value ?? "").trim();
const selectionKey = (row) => [row?.round, row?.game_no ?? row?.["게임번호"],
  row?.market, row?.market_label ?? row?.label ?? "", row?.sel ?? row?.["선택"]]
  .map(clean).join("|");
const lexical = (a, b) => a < b ? -1 : a > b ? 1 : 0;
const dateParts = (row) => {
  let kickoffText = clean(row?.kickoff_at);
  // Legacy naive timestamps mean KST, regardless of the browser's time zone.
  if (/^\d{4}-\d{2}-\d{2}[T ]\d{2}:\d{2}(?::\d{2}(?:\.\d+)?)?$/.test(kickoffText)) {
    kickoffText = `${kickoffText.replace(" ", "T")}+09:00`;
  }
  const kickoff = Date.parse(kickoffText);
  const date = clean(row?.date).match(/^(\d{2})\.(\d{2})/);
  const iso = Number.isFinite(kickoff)
    ? new Date(kickoff + 9 * 3600000).toISOString().slice(0, 10)
    : null;
  return { year: iso?.slice(0, 4) || clean(row?.year),
    monthDay: iso?.slice(5) || (date ? `${date[1]}-${date[2]}` : null) };
};
const leagueKey = (row, knownYears) => {
  const { year, monthDay } = dateParts(row);
  const known = knownYears.get(monthDay);
  const inferredYear = known?.size === 1 ? [...known][0] : "undated";
  const day = monthDay ? `${year || inferredYear}-${monthDay}` : "undated";
  return `${day}|${clean(row?.league) || "리그 미분류"}`;
};

export function compareDailyValue(a, b) {
  const left = dailyValueMetrics(a);
  const right = dailyValueMetrics(b);
  return (right.comparison_return ?? -Infinity) - (left.comparison_return ?? -Infinity) ||
    (right.expected_return ?? -Infinity) - (left.expected_return ?? -Infinity) ||
    (right.probability ?? -Infinity) - (left.probability ?? -Infinity) ||
    lexical(clean(a?.kickoff_at || a?.date), clean(b?.kickoff_at || b?.date)) ||
    lexical(selectionKey(a), selectionKey(b));
}

/** Same policy is persisted with the DB artifact by src/daily_value.py. */
export function dailyValueDecisions(candidates = []) {
  // Old candidates may have a date but no year; don't give the same KST day
  // two league quotas just because another candidate has an ISO timestamp.
  const knownYears = new Map();
  candidates.forEach((row) => {
    const { year, monthDay } = dateParts(row);
    if (!year || !monthDay) return;
    if (!knownYears.has(monthDay)) knownYears.set(monthDay, new Set());
    knownYears.get(monthDay).add(year);
  });
  const eligible = new Set(eligibleFinalSelections(candidates));
  const rows = candidates.map((selection) => {
    const metrics = dailyValueMetrics(selection);
    const odds = number(selection?.odds ?? selection?.["배당"]);
    const reason = !eligible.has(selection) || odds == null || odds <= 1 ? "safety"
      : metrics.probability == null ? "invalid"
        : metrics.probability < MIN_HIT ? "hit_floor"
          : !metrics.qualifies ? "return_floor" : "rank";
    return { selection, ...metrics, recommended: false, league_rank: null, reason_code: reason };
  });
  const groups = new Map();
  rows.filter((row) => row.reason_code === "rank").forEach((row) => {
    const key = leagueKey(row.selection, knownYears);
    if (!groups.has(key)) groups.set(key, []);
    groups.get(key).push(row);
  });
  groups.forEach((group) => {
    const primary = group.filter((row) => recommendationPriority(row.selection) === 1);
    const pool = primary.length ? primary : group;
    if (primary.length) group.filter((row) => !primary.includes(row))
      .forEach((row) => { row.reason_code = "fallback"; });
    pool.sort((a, b) => compareDailyValue(a.selection, b.selection)).forEach((row, index) => {
      row.league_rank = index + 1;
      const extra = row.validated_interval && row.comparison_return > 0;
      row.recommended = index < BASE_PER_LEAGUE || extra;
      row.reason_code = index < BASE_PER_LEAGUE ? "base" : extra ? "validated_extra" : "rank";
    });
  });
  return rows;
}
